pkg/cloudinfo/alibaba/cloudinfo.go

```diff
@@ -230,132 +198,303 @@ func (e *AlibabaInfoer) GetAttributeValues(ctx context.Context, service, attribu
+	request.Version = "2014-05-26"
+	request.ApiName = "DescribeZones"
+	request.QueryParams["RegionId"] = region
+
+	describeZones, err := a.client.ProcessCommonRequest(request)
+	if err != nil {
+		return nil, emperror.Wrap(err, "DescribeZones API call problem")
+	}
+
+	response := &ecs.DescribeZonesResponse{}
+
+	err = json.Unmarshal(describeZones.BaseResponse.GetHttpContentBytes(), response)
+	if err != nil {
+		return nil, err
+	}
+
+	return response.Zones.Zone, nil
+}
+
 // GetProducts retrieves the available virtual machines based on the arguments provided
-func (e *AlibabaInfoer) GetProducts(ctx context.Context, service, regionId string) ([]cloudinfo.VmInfo, error) {
+func (a *AlibabaInfoer) GetProducts(ctx context.Context, service, regionId string) ([]cloudinfo.VmInfo, error) {
 	log := logger.Extract(ctx)
 	log.Debug("getting product info")
-	var vms []cloudinfo.VmInfo
+	vms := make([]cloudinfo.VmInfo, 0)
 
-	request := ecs.CreateDescribeInstanceTypesRequest()
+	instanceTypes, err := a.getInstanceTypes()
+	if err != nil {
+		return nil, err
+	}
 
-	vmSizes, err := e.ecsClient.DescribeInstanceTypes(request)
+	availableZones, err := a.getZones(regionId)
 	if err != nil {
 		return nil, err
 	}
 
-	dataFromJson, err := e.priceRetriever.getOnDemandPrice(viper.GetString(priceInfoUrl))
+	for _, instanceType := range instanceTypes {
+		zones := make([]string, 0)
+		for _, zone := range availableZones {
+			for _, resourcesInfo := range zone.AvailableResources.ResourcesInfo {
+				for _, availableInstanceType := range resourcesInfo.InstanceTypes.SupportedInstanceType {
+					if availableInstanceType == instanceType.InstanceTypeId {
+						zones = append(zones, zone.ZoneId)
+						break
+					}
+				}
+			}
+		}
+		if len(zones) > 0 {
+
+			ntwMapper := newAlibabaNetworkMapper()
+			ntwPerf := fmt.Sprintf("%.1f Gbit/s", float64(instanceType.InstanceBandwidthRx)/1024000)
+			ntwPerfCat, err := ntwMapper.MapNetworkPerf(ntwPerf)
+			if err != nil {
+				log.Debug("could not get network performance category")
+			}
+
+			vms = append(vms, cloudinfo.VmInfo{
+				Type:       instanceType.InstanceTypeId,
+				Cpus:       float64(instanceType.CpuCoreCount),
+				Mem:        instanceType.MemorySize,
+				Gpus:       float64(instanceType.GPUAmount),
+				NtwPerf:    ntwPerf,
+				NtwPerfCat: ntwPerfCat,
+				Zones:      zones,
+				Attributes: cloudinfo.Attributes(fmt.Sprint(instanceType.CpuCoreCount), fmt.Sprint(instanceType.MemorySize), ntwPerfCat),
+			})
+		}
+	}
+
+	virtualMachines, err := a.getOnDemandPrice(vms, regionId)
 	if err != nil {
 		return nil, err
 	}
 
-	zoneRequest := ecs.CreateDescribeZonesRequest()
-	request.RegionId = regionId
+	log.Debug("found vms", map[string]interface{}{"vms": fmt.Sprintf("%v", virtualMachines)})
+	return virtualMachines, nil
+}
 
-	zoneResponse, err := e.ecsClient.DescribeZones(zoneRequest)
+func (a *AlibabaInfoer) getInstanceTypes() ([]ecs.InstanceType, error) {
+	request := requests.NewCommonRequest()
```

Comment:
let's discuss the code organization in methods offline: accessing the cloud should be separated from processing the response